### PR TITLE
Open fewer simultaneous filehandles in AlignmentSummaryV2

### PR DESCRIPTION
**Pull request by tmooney**
AlignmentSummaryV2 splits the BED into one per chromosome. For the human reference, this isn't too many, but for other combined references this may exceed the maximum number of allowed filehandles.  This PR closes the previous ones if it would open a 101st and lets them be reopened as needed.

### CODE REVIEW

File: lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm [hunk 1]
```diff
@@ -577,8 +577,13 @@ sub _prepare_split_sorted_bed_files {
         my $chrom = $1;
 
         unless ($open_handles{$chrom}) {
+            if(scalar keys %open_handles >= 100) {
+                $_->close foreach values %open_handles;
+                %open_handles = ();
```
Comment (nnutter) [inline lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm]: Does this break next for loop where it is "Now sort each file"?

File: lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm [hunk 2]
```diff
@@ -577,8 +577,13 @@ sub _prepare_split_sorted_bed_files {
         my $chrom = $1;
 
         unless ($open_handles{$chrom}) {
+            if(scalar keys %open_handles >= 100) {
+                $_->close foreach values %open_handles;
```
Comment (nnutter) [inline lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm]: Maybe it should just cache a filehandle for consecutive writes for the same chromosome but close and open anytime the chromosome changes?  I don't know what the data looks like.

File: lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm (hunk 1 above)
Comment (tmooney) [inline lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm]: Quite right you are!  I'll have to change this.

File: lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm (hunk 2 above)
Comment (tmooney) [inline lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm]: A lot of our input files are not sorted at all and jump around from chromosome to chromosome.  I had also thought about sorting the file in advance of this loop, so we knew we only needed one filehandle at a time.

File: lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm (hunk 1 above)
Comment (tmooney) [inline lib/perl/Genome/Model/Tools/BioSamtools/AlignmentSummaryV2.pm]: I've added a separate hash to track all the chromosomes we've seen previously.

**Comment by nnutter** [comment]
:+1:

**Comment by brummett** [comment]
+1

If you're interested, the FileCache module basically does the same thing and it comes with Perl.

**Comment by nnutter** [comment]
[FileCache](http://perldoc.perl.org/5.8.9/FileCache.html)++

**Comment by tmooney** [comment]
`FileCache` is quite odd:
> The module functionality relies on symbolic references, so things will break under 'use strict' unless 'no strict "refs"' is also specified.

This code works:
```perl
package This::Is::A::Test;
use strict;
use warnings;
use FileCache maxopen => 16;

my $filename = "/tmp/test";
cacheout($filename) or die "Can't open file $filename for writing: $!";
my $fh = __PACKAGE__ . '::' . $filename;
no strict 'refs';
print $fh "This is a test.\n";
```

This code does not:
```perl
package This::Is::A::Test;
use strict;
use warnings;
use FileCache maxopen => 16;

my $filename = "/tmp/test";
my $fh = cacheout($filename) or die "Can't open file $filename for writing: $!";
no strict 'refs';
print $fh "This is a test.\n";
```
(This example produces `print() on unopened filehandle /tmp/test at - line 9.` and thus no output in the file.)

**Comment by nnutter** [comment]
[FileCache](http://perldoc.perl.org/5.8.9/FileCache.html)--

**Comment by acoffman** [comment]
Personally, I think the extra clarity that comes from using FileCache isn't worth turning strict refs off, but 
I'm fine with either approach.

+1 either way.

**Comment by brummett** [comment]
Looks like FileCache is a relic from Perl4 or early Perl5.  Forget I even suggested it...

My previous +1 still stands

**Pull request merged**